server/src/main/java/io/deephaven/server/partitionedtable/PartitionedTableServiceGrpcImpl.java

```diff
@@ -0,0 +1,126 @@
+import io.deephaven.io.logger.Logger;
+import io.deephaven.proto.backplane.grpc.ExportedTableCreationResponse;
+import io.deephaven.proto.backplane.grpc.GetTableRequest;
+import io.deephaven.proto.backplane.grpc.MergeRequest;
+import io.deephaven.proto.backplane.grpc.PartitionByRequest;
+import io.deephaven.proto.backplane.grpc.PartitionByResponse;
+import io.deephaven.proto.backplane.grpc.PartitionedTableServiceGrpc;
+import io.deephaven.server.session.SessionService;
+import io.deephaven.server.session.SessionState;
+import io.deephaven.server.session.TicketRouter;
+import io.grpc.stub.StreamObserver;
+
+import javax.inject.Inject;
+
+import static io.deephaven.extensions.barrage.util.GrpcUtil.safelyExecute;
+
+public class PartitionedTableServiceGrpcImpl extends PartitionedTableServiceGrpc.PartitionedTableServiceImplBase {
+    private static final Logger log = LoggerFactory.getLogger(PartitionedTableServiceGrpcImpl.class);
+
+    private final TicketRouter ticketRouter;
+    private final SessionService sessionService;
+
+    @Inject
+    public PartitionedTableServiceGrpcImpl(TicketRouter ticketRouter, SessionService sessionService) {
+        this.ticketRouter = ticketRouter;
+        this.sessionService = sessionService;
+    }
+
+    @Override
+    public void partitionBy(PartitionByRequest request, StreamObserver<PartitionByResponse> responseObserver) {
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            final SessionState session = sessionService.getCurrentSession();
+
+            SessionState.ExportObject<Table> targetTable =
+                    ticketRouter.resolve(session, request.getTableId(), "tableId");
+
+            session.newExport(request.getResultId(), "resultId")
+                    .require(targetTable)
+                    .onError(responseObserver)
+                    .submit(() -> {
+                        PartitionedTable partitionedTable = targetTable.get().partitionBy(request.getDropKeys(),
+                                request.getKeyColumnNamesList().toArray(String[]::new));
+                        safelyExecute(() -> {
+                            responseObserver.onNext(PartitionByResponse.getDefaultInstance());
+                            responseObserver.onCompleted();
+                        });
+                        return partitionedTable;
+                    });
+
+        });
+    }
+
+    @Override
+    public void merge(MergeRequest request, StreamObserver<ExportedTableCreationResponse> responseObserver) {
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            final SessionState session = sessionService.getCurrentSession();
+
+            SessionState.ExportObject<PartitionedTable> partitionedTable =
+                    ticketRouter.resolve(session, request.getPartitionedTable(), "partitionedTable");
+
+            session.newExport(request.getResultId(), "resultId")
+                    .require(partitionedTable)
```

Comment:
We need to run this merge under the UGP shared lock if `PartitionedTable.table().isRefreshing()`.
Can we *conditionally* add `                    .requiresSerialQueue()` ?

Reply:
Can't conditionally add it, but we can conditionally wrap the contents of the submit

Reply:
After our discussion, I think we agree that what we want to do is conditionally acquire the shared lock, but continue running this on the concurrent queue.